sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala

```diff
@@ -236,169 +232,143 @@ object GpuShuffledHashJoinExec extends Arm {
+          (Left(singleBuildCb), bufferedStreamIter)
+
+        } else { // Other cases without optimization
+          val safeIter = GpuSubPartitionHashJoin.safeIteratorFromSeq(Seq(firstBuildBatch)) ++
+            coalesceBuiltIter
+          val gpuBuildIter = if (isBuildSerialized) {
+            // batches on host, move them to GPU
+            new GpuShuffleCoalesceIterator(safeIter.asInstanceOf[Iterator[HostConcatResult]],
+              buildTypes, coalesceMetrics)
+          } else { // batches already on GPU
+            safeIter.asInstanceOf[Iterator[ColumnarBatch]]
+          }
+
+          val buildRet = if (needSingleBuildBatch) {
+            val singleBuildCb = getAsSingleBatch(gpuBuildIter, buildOutput,
+              hasMultipleBatches, buildGoal, spillCallback, coalesceMetrics)
+            Left(singleBuildCb)
+          } else { // this is for sub-partitioning
+            Right(new CollectTimeIterator("hash join build", gpuBuildIter, buildTime))
           }
+          // add the time it took to fetch that first build batch
+          buildTime += System.nanoTime() - startTime
+          (buildRet, streamIter)
         }
+      } else {
+        // build is empty
+        (Left(GpuColumnVector.emptyBatchFromTypes(buildTypes)), streamIter)
       }
     }
   }
 
-  private def getBuildBatchFromUnfinished(
-      buildGoal: CoalesceSizeGoal,
-      iterWithPrior: Iterator[HostConcatResult],
-      buildOutput: Seq[Attribute],
-      spillCallback: SpillCallback,
-      coalesceMetricsMap: Map[String, GpuMetric]): ColumnarBatch = {
-    // In the fallback case we build the same iterator chain that the Spark plan
-    // would have produced:
-    //   GpuCoalesceIterator(GpuShuffleCoalesceIterator(shuffled build side))
-    // This allows us to make the shuffle batches spillable in case we have a large,
-    // build-side table, as `RequireSingleBatch` is virtually no limit, and we
-    // know we are now above `hostTargetBatchSize` (which is 2GB by default)
-    val dataTypes = buildOutput.map(_.dataType).toArray
-      val shuffleCoalesce = new GpuShuffleCoalesceIterator(
-        iterWithPrior,
-        dataTypes,
-        coalesceMetricsMap)
-    val res = ConcatAndConsumeAll.getSingleBatchWithVerification(
-        new GpuCoalesceIterator(shuffleCoalesce,
-          dataTypes, buildGoal,
-          NoopMetric, NoopMetric, NoopMetric, NoopMetric, NoopMetric,
-          coalesceMetricsMap(GpuMetric.CONCAT_TIME),
-          coalesceMetricsMap(GpuMetric.OP_TIME),
-          coalesceMetricsMap(GpuMetric.PEAK_DEVICE_MEMORY),
-          spillCallback,
-          "build batch"),
-        buildOutput)
-      res
+  /** Only accepts a HostConcatResult or a ColumnarBatch as input */
+  private def getBatchSize(maybeBatch: AnyRef): Long = maybeBatch match {
+    case batch: ColumnarBatch => GpuColumnVector.getTotalDeviceMemoryUsed(batch)
+    case hostBatch: HostConcatResult => hostBatch.getTableHeader().getDataLen()
```

Comment:
Can we put in a case for anything else that throws an exception. I know that Scala will add it in for us, but it also outputs a warning when it does this.

Reply:
sure

Reply:
done